### PR TITLE
feat(platform): Add support for JetBrains Space platform

### PR DESCRIPTION
**Pull request by vooft**
## Changes

Hi All,

I implemented integration with the JetBrains Space platform.
Features supported:
- Autodiscover
- Creating PRs
- Finding closed ones
- Rebasing PRs (manual only via adding `rebase!` to the title)
- Adding/removing comments
- Changing PR body
- Adding reviewers
- Adding authors
- Auto-merging if checks have passed (not available in free plan, but could be tested on trial)

Not everything is covered with tests yet, but, say, 70% of the new code. I would like to get some feedback on my code, please, before proceeding with the remaining tests. My background is in JVM world and, perhaps, some things were not implemented idiomatically.

Unfortunately, Space doesn't have public repositories, but I'm happy to give access for interested to my test environment.

Minimal config for testing would look like this:
```js
module.exports = {
  token: '<your-personal-token>',
  repositories: ['main/my-repo'], // optional, goes like <project-key>/<repo-name>, default project key is "MAIN", but it seems that server doesn't care about casing
  platform: 'space',
  endpoint: '<your-org>.jetbrains.space',
};
```

## Context

I was using Renovate for long time and wanted to contribute in some way, as well as to play with JetBrains Space.

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/workers/repository/onboarding/branch/check.ts
Comment (rarkins) [inline lib/workers/repository/onboarding/branch/check.ts:63]: Needs reverting

File: lib/workers/repository/update/pr/index.ts
Comment (rarkins) [inline lib/workers/repository/update/pr/index.ts:446]: Is this actually helpful to the user? We don't intend for these hashes to be meaningful to anyone

File: lib/modules/platform/space/utils.ts
Comment (rarkins) [inline lib/modules/platform/space/utils.ts:112]: Is this a kind of "hack" to determine whether a closed PR was merged or not? Would be good to document it so that others don't need to guess in future

File: lib/modules/platform/space/index.ts
Comment (rarkins) [inline lib/modules/platform/space/index.ts:108]: Just a general note that logging like this will need to be lowered or removed before merge

**Review by rarkins — CHANGES_REQUESTED** [review]
Thanks for this PR! The most important point would be to adopt Zod for validating all responses from the Space API. Right now you are casting `any` to typed values, which is risky for runtime faults. Instead we want to use Zod for all new code like this: https://github.com/renovatebot/renovate/blob/main/docs/development/zod.md

File: lib/modules/platform/space/index.ts
Comment (vooft) [reply lib/modules/platform/space/index.ts:108]: Right, I'll clean up most of the debug logs, thanks

File: lib/workers/repository/update/pr/index.ts
Comment (vooft) [reply lib/workers/repository/update/pr/index.ts]: No, I added it when was debugging, removed

File: lib/modules/platform/space/index.ts
Comment (vooft) [reply lib/modules/platform/space/index.ts]: Replaced most of logs with trace or simply removed, kept a couple that might be useful for general user debugging

File: lib/modules/platform/space/utils.ts
Comment (vooft) [reply lib/modules/platform/space/utils.ts:105]: Added a comment explaining why need this flag

**Comment by vooft** [comment]
@rarkins thank you for checking it out! Will read about ZOD and make required changes.

File: lib/modules/platform/space/client/code-review-read.spec.ts
Comment (HonkingGoose) [inline lib/modules/platform/space/client/code-review-read.spec.ts:14]: I think something like `jetbrainsSpaceEndpointUrl` is easier to understand.

File: lib/config/presets/local/index.ts
Comment (HonkingGoose) [inline lib/config/presets/local/index.ts:26]: There's a typo in the PR title, please fix it. 🙂

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

File: lib/config/presets/local/index.ts
Comment (viceice) [inline lib/config/presets/local/index.ts:29]:
```suggestion
  space: local,
```

loading presets from local platform

File: lib/modules/platform/space/types.ts
Comment (viceice) [inline lib/modules/platform/space/types.ts:1]: Use zod schema for types

https://github.com/renovatebot/renovate/blob/main/docs/development/zod.md#L1

**Review by viceice — CHANGES_REQUESTED** [review]
Spaces is deprecated[^1]!

@rarkins @secustor Should we add a deprecated platform? Support will be fully stoped  on 2025-05-31.

They currently have a new platform in preview `SpaceCode`.
I'm sure it's totally different.

[^1]: https://blog.jetbrains.com/space/2024/05/27/the-future-of-space/

**Comment by vooft** [comment]
@viceice that's very unfortunate lol, I chose the perfect time to start building this integration :)
I imagine they will probably reuse part of the code from Space in the new platform, so potentially the api will be similar.

**Comment by viceice** [comment]
> @viceice that's very unfortunate lol, I chose the perfect time to start building this integration :)
> I imagine they will probably reuse part of the code from Space in the new platform, so potentially the api will be similar.

it's already in preview, so you can check it.

**Comment by vooft** [comment]
The API feels mostly the same, looks like they removed chats, but some chat-related APIs are still there.
I will probably pause for now until they release it to public.

**Comment by HonkingGoose** [comment]
> They currently have a new platform in preview SpaceCode.
> I'm sure it's totally different.

I found some more info about the differences between the old Space and the upcoming SpaceCode. The [Jetbrain docs, migrate from Space to SpaceCode, what will not stay in SpaceCode](https://www.jetbrains.com/help/space/migrate-from-space-to-spacecode.html#what-will-not-stay-in-spacecode) says:

> **What will NOT stay in SpaceCode**﻿
> 
> - [Automation](https://www.jetbrains.com/help/space/migrate-from-space-to-spacecode.html#automation-ci-cd). You will be able to convert and transfer your Automation jobs to JetBrains TeamCity and continue using them for your CI/CD needs.
> - [Issues](https://www.jetbrains.com/help/space/migrate-from-space-to-spacecode.html#issues-issue-tracker). You will be able to move your issues to an alternative planning tool — JetBrains YouTrack.
> - [Documents](https://www.jetbrains.com/help/space/migrate-from-space-to-spacecode.html#documents-and-blogs). Project documents can be exported to YouTrack Knowledge Base.
> - [Checklists](https://www.jetbrains.com/help/space/migrate-from-space-to-spacecode.html#blogs). Can be extracted via HTTP API.
> - [Blog posts](https://www.jetbrains.com/help/space/migrate-from-space-to-spacecode.html#blogs). Can be extracted via HTTP API.
> - [Chats](https://www.jetbrains.com/help/space/migrate-from-space-to-spacecode.html#chats-and-messages). Messages can be retrieved in batch via HTTP API.
> - [Packages](https://www.jetbrains.com/help/space/migrate-from-space-to-spacecode.html#packages). Package repository source data can be accessed and retrieved via HTTP API.
> - [Calendars](https://www.jetbrains.com/help/space/migrate-from-space-to-spacecode.html#calendars) and calendar events such as Meetings, Absences, Holidays. Personal calendars can be exported to .ICS files.
> - To-do lists.
> - Locations.
> - Work schedule.
> - Availability status.

I'll let the maintainers decide if it's worth it to support a now deprecated platform.

**Comment by HonkingGoose** [comment]
So what are we going to do?

- leave PR open and fix it up when the _newest_ JetBrains thing comes out?
- add note to our docs about the _short_ support window, merge PR, and later drop Renovate support in a breaking release when JetBrains drops support
- close the PR and thank the PR author for their work

**Comment by vooft** [comment]
@HonkingGoose I'm perfectly happy to continue this integration once JetBrains fully release the new platform. I can open a new PR once I have something working.

**Comment by HonkingGoose** [comment]
Thank you @vooft for trying to get this feature landed. ❤️

**Pull request closed without merging**
